actions/charts/longhorn.go

```diff
@@ -0,0 +1,137 @@
+	"github.com/rancher/shepherd/pkg/wait"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/watch"
+)
+
+var (
+	longhornChartName     = "longhorn"
+	enableDeletionSetting = map[string]any{
+		"defaultSettings": map[string]any{
+			"deletingConfirmationFlag": true,
+		},
+	}
+)
+
+type LonghornGlobalSettingPut struct {
+	Links map[string]string `json:"links"`
+	ID    string            `json:"id"`
+	Name  string            `json:"name"`
+	Type  string            `json:"type"`
+	Value string            `json:"value"`
+}
+
+// InstallLonghornChart installs ther Longhorn chart on the cluster related to the received catalog client object.
+// Enable the wait parameter if you wish for this function to wait for all deployments and daemonsets to be ready before returning.
+func InstallLonghornChart(client *rancher.Client, payload PayloadOpts, values map[string]interface{}) error {
+	catalogClient, err := client.GetClusterCatalogClient(payload.Cluster.ID)
+	if err != nil {
+		return err
+	}
+
+	// If no specific value for setting deletingConfirmationFlag was provided, default to have it enabled so cleanup works as extected.
+	if values == nil {
+		values = enableDeletionSetting
+	} else {
+		defaultSettings, ok := values["defaultSettings"].(map[string]any)
+		if !ok {
+			return errors.New(`Provided values map has invalid value for "defaultSettings"`)
+		}
+
+		_, ok = defaultSettings["deletingConfirmationFlag"]
+		if !ok {
+			defaultSettings["deletingConfirmationFlag"] = true
+		}
+	}
+
+	chartInstalls := []types.ChartInstall{
+		*NewChartInstall(longhornChartName+"-crd", payload.Version, payload.Cluster.ID, payload.Cluster.Name, payload.Host, catalog.RancherChartRepo, payload.ProjectID, payload.DefaultRegistry, nil),
+		*NewChartInstall(longhornChartName, payload.Version, payload.Cluster.ID, payload.Cluster.Name, payload.Host, catalog.RancherChartRepo, payload.ProjectID, payload.DefaultRegistry, values),
+	}
+
+	chartInstallAction := NewChartInstallAction(payload.Namespace, payload.ProjectID, chartInstalls)
+
+	err = catalogClient.InstallChart(chartInstallAction, catalog.RancherChartRepo)
+	if err != nil {
+		return err
+	}
+
+	client.Session.RegisterCleanupFunc(func() error {
+		chart, err := shepherdCharts.GetChartStatus(client, payload.Cluster.ID, payload.Namespace, longhornChartName)
+		if err != nil || !chart.IsAlreadyInstalled {
+			return err
+		} else {
+			return UninstallLonghornChart(client, payload.Namespace, payload.Cluster.ID, payload.Host)
+		}
+	})
+
+	err = shepherdCharts.WaitChartInstall(catalogClient, payload.Namespace, longhornChartName)
+	if err != nil {
+		return err
+	}
+
+	err = shepherdCharts.WatchAndWaitDeployments(client, payload.Cluster.ID, payload.Namespace, metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	err = shepherdCharts.WatchAndWaitDaemonSets(client, payload.Cluster.ID, payload.Namespace, metav1.ListOptions{})
+	return err
```

Comment:
Is it possible to add label selectors that specifically target Longhorn workloads to the metav1.ListOptions instead of being blank? It's possible, though unlikely, that other workloads could be running in the namespace